Implement Status API Endpoint in Pet and User

anishprasad01 opened the pull request:
Implementing Status API Endpoints for Pets and Users to get a single pet's status, or all 3 of a user's pets
Partially Blocked by #58 

Fix #50

anishprasad01 committed:
Created StatusRequest type
Created PetStatus type
Added new endpoint to Pet (DOES NOT WORK YET)

anishprasad01 committed:
Created StatusRequest type
Created PetStatus type
Added new endpoint to Pet (DOES NOT WORK YET)

anishprasad01 committed:
Created StatusRequest type
Created PetStatus type
Added new endpoint to Pet (DOES NOT WORK YET)

anishprasad01 committed:
Created StatusRequest type
Created PetStatus type
Added new endpoint to Pet (DOES NOT WORK YET)

anishprasad01 committed:
Makes GetPetStatus temp compile

anishprasad01 committed:
Makes GetPetStatus temp compile

anishprasad01 committed:
Makes GetPetStatus temp compile

anishprasad01 committed:
Implemented Status in PetController

anishprasad01 committed:
Implemented Status in PetController

anishprasad01 committed:
Implemented Status in PetController

anishprasad01 committed:
Temporary return in PetService for Status

anishprasad01 committed:
Temporary return in PetService for Status

anishprasad01 committed:
Temporary return in PetService for Status

anishprasad01 committed:
Temporary return in PetService for Status

anishprasad01 committed:
Pet Status endpoint works.

Created a new PetStatus type
Added GetPetStatus to PetService
added Pet Status endpoint to PetController

anishprasad01 committed:
Pet Status endpoint works.

Created a new PetStatus type
Added GetPetStatus to PetService
added Pet Status endpoint to PetController

anishprasad01 committed:
Pet Status endpoint works.

Created a new PetStatus type
Added GetPetStatus to PetService
added Pet Status endpoint to PetController

anishprasad01 committed:
Pet Status endpoint works.

Created a new PetStatus type
Added GetPetStatus to PetService
added Pet Status endpoint to PetController

anishprasad01 committed:
Fixed CodeFactor error

anishprasad01 committed:
Fixed CodeFactor error

anishprasad01 committed:
Fixed CodeFactor error

anishprasad01 committed:
Fixed CodeFactor error

Chris-Johnston commented on PetGame.Core/Models/PetStatus.cs:
```diff
@@ -0,0 +1,38 @@
+        /// </summary>
+        [JsonProperty]
+        [JsonConverter(typeof(IsoDateTimeConverter))]
+        public DateTime TimeOfNextAction { get; set; } = DateTime.Now.AddMinutes(5);
```
This is incorrect, actions like racing, feeding, and training should each track when they can be done next separately.

Chris-Johnston commented on PetGame/Controllers/PetController.cs:
```diff
@@ -50,6 +50,34 @@ public IActionResult Get(ulong id)
+        /// JSON of PetStatus object
+        /// </returns>
+        [HttpGet("{id}/status"), AllowAnonymous]
+        public IActionResult Get(StatusRequest Request)
```
You are on the right track, but we shouldn't box the ID into it's own object. The names of parameters have to match that in the route definition, which is (probably) why you were getting the conflicting issue as described in StatusRequest.
```suggestion
        public IActionResult Get(ulong id)
```

Chris-Johnston commented on PetGame/Models/StatusRequest.cs:
```diff
@@ -0,0 +1,11 @@
+    /// Used to encapsulate the id in order to avoid issues with identical
+    /// function signatures in PetController and UserController
+    /// </summary>
+    public class StatusRequest
```
This class is no longer necessary. Also, in the future, I would pick a name that is more descriptive.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                //gets all activities from the last 2 hours for specified pet
+                cmd.CommandText = @"SELECT Activity.ActivityId, Activity.PetId, Activity.Timestamp, Activity.Type 
+                                    FROM Activity 
+                                    WHERE Activity.Timestamp > DATEADD(HOUR, -2, GETDATE()) AND Pet.PetId = @PetId;";
```
let's make `-2` into a parameter, and then add that as a constant in this class. I think we should target something more like 12 hours, checking the game every 2 would be a nightmare to play.

Chris-Johnston replied:
But, making this configurable means we can use whatever we want, so it's easier to test.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                        if (Type.Equals("Training"))
+                        {
+                            ActivityTypeInput = ActivityType.Training;
+                        }
```
This is not the way to read the activity from the database. The `Type` column in the database is a `char(1)`. Take a look how this is done in `ActivityService`: https://github.com/Chris-Johnston/PGTRTKD3000DGOTY/blob/master/PetGame/Services/ActivityService.cs#L69

... also wait why aren't you just using that method anyways? 

https://github.com/Chris-Johnston/PGTRTKD3000DGOTY/blob/master/PetGame/Services/ActivityService.cs#L30

anishprasad01 replied:
I tried reading it as a char to begin with, but that did not work. I'll try calling the ActivityService function instead.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                else
+                {
+                    //check the type of activity and subtract percentage accordingly
+                    foreach (Activity Activity in LastTwoHoursActivities)
```
Rename `LastTwoHoursActivities` to something that is still descriptive if we change that amount of time.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                    foreach (Activity Activity in LastTwoHoursActivities)
+                    {
+                        //if only one hour has passed, subtract 5%
+                        if ((Activity.Timestamp.Hour + 1).Equals(DateTime.Now))
```
I would recommend using `TimeSpan.Add` instead here.

Chris-Johnston replied:
Also, this isn't Java, we can `==` objects without issue.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                            hungerPercentage -= 0.05;
+                        }
+                        //if a traing activity has occured, subtract 7%
+                        else if (Activity.Type.Equals(ActivityType.Training))
```
switch/case please, we are going to have more than just these activities

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                            hungerPercentage -= 0.07;
+                        }
+                        //if a race has occured, subtract 15%
+                        else if (Activity.Type.Equals(ActivityType.Default))
```
`ActivityType.Default` is not a race. This enum needs to be updated with the rest of the types (so I guess I'll do that tonight)

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                        //{
+                        //    //bad default case?
+                        //    hungerPercentage -= 0.00;
+                        //}
```
Remove commented out code.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+
+                //multiply 10% by the number of hours and the hunger percentage to get the
+                //happiness percentage
+                happinessPercentage = (HoursSinceLastActivity * 0.10 * hungerPercentage);
```
Magic numbers = bad. Please make these constants.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                }//end
+
+                //get the last activity
+                Activity LastActivity = LastTwoHoursActivities[LastTwoHoursActivities.Count - 1];
```
This will index out of bounds if the list is empty.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -196,5 +196,118 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                PetStatus ret = new PetStatus() { Pet = toReturn, Hunger = hungerPercentage, Happiness = happinessPercentage };
+
+                //return the new object
+                return ret;
```
You can `return new PetStatus() { ... };` instead

Chris-Johnston commented:
Oh oops, didn't see the WIP tag.

anishprasad01 committed:
Implement Status endpoint for user

anishprasad01 committed:
Implement Status endpoint for user

anishprasad01 committed:
Implement Status endpoint for user

anishprasad01 committed:
Implement Status endpoint for user

anishprasad01 committed:
added no-pet check

added code to return the statuses as Json in UserController

anishprasad01 committed:
added no-pet check

added code to return the statuses as Json in UserController

anishprasad01 committed:
added no-pet check

added code to return the statuses as Json in UserController

anishprasad01 committed:
added no-pet check

added code to return the statuses as Json in UserController

anishprasad01 commented:
No, that's cool. I'd rather have feedback now than when I'm totally done

anishprasad01 committed:
Changed all numbers used in GetPetStatus to constants

Changed # of hours in query to get activities to a parameter and constant

Removed StatusRequest class and changed the function signature in PetController

Misc. Code Cleanup

anishprasad01 committed:
Changed all numbers used in GetPetStatus to constants

Changed # of hours in query to get activities to a parameter and constant

Removed StatusRequest class and changed the function signature in PetController

Misc. Code Cleanup

anishprasad01 committed:
Changed all numbers used in GetPetStatus to constants

Changed # of hours in query to get activities to a parameter and constant

Removed StatusRequest class and changed the function signature in PetController

Misc. Code Cleanup

anishprasad01 committed:
Changed all numbers used in GetPetStatus to constants

Changed # of hours in query to get activities to a parameter and constant

Removed StatusRequest class and changed the function signature in PetController

Misc. Code Cleanup

anishprasad01 committed:
Moved position of constants on PetService to top of the class

Fixed an error in UserController caused by removal of StatusRequest

anishprasad01 committed:
Moved position of constants on PetService to top of the class

Fixed an error in UserController caused by removal of StatusRequest

anishprasad01 committed:
Moved position of constants on PetService to top of the class

Fixed an error in UserController caused by removal of StatusRequest

anishprasad01 committed:
Moved position of constants on PetService to top of the class

Fixed an error in UserController caused by removal of StatusRequest

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -229,11 +236,12 @@ public PetStatus GetPetStatus(ulong PetId)
-                                    FROM Activity 
-                                    WHERE Activity.Timestamp > DATEADD(HOUR, -2, GETDATE()) AND Pet.PetId = @PetId;";
+                                    FROM Activity, Pet 
+                                    WHERE Activity.Timestamp > DATEADD(HOUR, -@Hours, GETDATE()) AND Pet.PetId = @PetId;";
```
Just make this `@Hours`

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -229,11 +236,12 @@ public PetStatus GetPetStatus(ulong PetId)
                 //specify PetId
-                cmd.Parameters.AddWithValue("PetId", PetId);
+                cmd.Parameters.AddWithValue("PetId", $"{PetId}");
+                cmd.Parameters.AddWithValue("Hours", HoursToCheck);
```
Missing the `@`

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -300,14 +303,46 @@ FROM Activity
+
+                cmd.CommandText = @"SELECT Pet.PetId FROM PET WHERE Pet.UserId = @UserId;";
+
+                cmd.Parameters.AddWithValue("UserId", $"{UserId}");
```
Missing `@`

anishprasad01 committed:
Modified the Constant for Hours in PetService to remove the - sign from the query

Added missing @ symbols in query parameter assignment code

anishprasad01 committed:
Modified the Constant for Hours in PetService to remove the - sign from the query

Added missing @ symbols in query parameter assignment code

anishprasad01 committed:
Modified the Constant for Hours in PetService to remove the - sign from the query

Added missing @ symbols in query parameter assignment code

anishprasad01 committed:
Added a check in case there are no activities in the interval of hours specified

anishprasad01 committed:
Added a check in case there are no activities in the interval of hours specified

anishprasad01 committed:
Added a check in case there are no activities in the interval of hours specified

anishprasad01 committed:
Fixed negative happiness bug

Added checks for negative hunger and negative happiness

Removed hunger value from happiness calculation

anishprasad01 committed:
Fixed negative happiness bug

Added checks for negative hunger and negative happiness

Removed hunger value from happiness calculation

anishprasad01 committed:
Fixed negative happiness bug

Added checks for negative hunger and negative happiness

Removed hunger value from happiness calculation

anishprasad01 committed:
Merge branch 'implement-status-api' of https://github.com/Chris-Johnston/PGTRTKD3000DGOTY into implement-status-api

anishprasad01 committed:
?

anishprasad01 committed:
?

anishprasad01 committed:
Removed Activity query for Status and replaced it with a call to ActivityService

anishprasad01 committed:
Removed Activity query for Status and replaced it with a call to ActivityService

anishprasad01 committed:
Merge branch 'implement-status-api' of https://github.com/Chris-Johnston/PGTRTKD3000DGOTY into implement-status-api

anishprasad01 committed:
Changing the conditions for Hunger level decrease and increase

anishprasad01 committed:
Modified Hunger decrease code to ensure that hunger will never be > 1 or < 0

anishprasad01 committed:
Modified Status to use a DateTime to specify the number of days to check. Added const variables for number of days and number of activities.

Added Feeding TIme, Race Time, and Training Time to Pet Status type

anishprasad01 committed:
Added check for invalid Happiness percentages.

Fixed Happiness always increasing bug.

Fixed non-nullable activitytype in GetActivity function

Chris-Johnston commented on PetGame/Services/ActivityService.cs:
```diff
@@ -53,7 +53,7 @@ FROM Activity
                 cmd.Parameters.AddWithValue("@Limit", limit);
                 // if type is null, then insert a null. this will disable filtering by type
-                cmd.Parameters.AddWithValue("@Type", (object) (char)type ?? DBNull.Value);
+                cmd.Parameters.AddWithValue("@Type", (object) (char)type.GetValueOrDefault() ?? DBNull.Value);
```
No, not `GetValueOrDefault()`, the `Value` property. https://docs.microsoft.com/en-us/dotnet/api/system.nullable-1.value?view=netframework-4.7.2#System_Nullable_1_Value
```suggestion
                cmd.Parameters.AddWithValue("@Type", (object) (type.HasValue ? (char)type.Value : DBNull.Value));
```

anishprasad01 replied:
It didn't like the `Value` property

anishprasad01 replied:
I will use the above, however.

anishprasad01 committed:
Fixed issue where Hunger never decreased

Added more comments

Renamed Get pet status fucntion to be more descriptive

anishprasad01 committed:
Merge branch 'master' into implement-status-api

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            DateTime ToCheck = DateTime.Now.Subtract(TimeSpan.FromDays(NumDaysToCheck));
+            int HoursToCheck = NumDaysToCheck * 24;
+
+            List<Activity> PastActivities = (List<Activity>)activityService.GetActivities(PetId, NumActivities, ToCheck, null);
```
You should not cast an `IEnumerable` into a `List`. Keep it as an `IEnumerable`.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            else
+            {
+                //check the type of activity and subtract percentage accordingly
+                foreach (Activity Activity in PastActivities)
```
No point in converting the `IEnumerable` into the list, since you are `foreach`-ing through it here.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                        else
+                        {
+                            hungerPercentage = 0;
+                        }
```
You repeat yourself a lot here, and arguably each of these actions could be broken into their own method.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            else
+            {
+                //get the last activity
+                Activity LastActivity = PastActivities[PastActivities.Count - 1];
```
The list of `PastActivities` is sorted by most recent to least recent. You'll always end up getting the activity that's the furthest away from the current time.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            else if (hungerPercentage > 1.0)
+            {
+                hungerPercentage = 1.0;
+            }
```
Again, you repeat yourself a lot here. I think you should implement a `double constrain(double value, double min, double max)` function, since you perform the same logic a few times here.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+
+                if (PetIdList.Count == 0)
+                {
+                    return null;
```
Why would this return null, and not an empty list? I don't think you should do that.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                    PetStatusList.Add(GetPetStatusById(Id));
+                }
+            }
+            return (PetStatusList);
```
Unnecessary parenthesis.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            {
+                var cmd = conn.CreateCommand();
+
+                cmd.CommandText = @"SELECT Pet.PetId FROM PET WHERE Pet.UserId = @UserId;";
```
The `Pet` table is not `PET`.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+
+                foreach (ulong Id in PetIdList)
+                {
+                    PetStatusList.Add(GetPetStatusById(Id));
```
You do this while already having a connection to the SQL database established. You really should do this outside of the `using` block.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +211,196 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                TimeOfNextAction = DateTime.Now.AddMinutes(MinutesTillNextAction)};
+        }//end of function
+
+        public IEnumerable<PetStatus> GetPetStatusList(ulong UserId)
```
Without any comments, I initially had no idea what this did.

anishprasad01 replied:
There are comments there now, right?

Chris-Johnston commented on PetGame/Controllers/UserController.cs:
```diff
@@ -52,19 +52,12 @@ public IActionResult GetWhoAmI()
-        [HttpPost("Pet")]
-        public IActionResult PostPet([FromForm, FromBody] CreatePetModel model)
+        [HttpGet("{id}/status")]
+        public IActionResult Get(ulong id)
```
Why is the `id` in here? You really should only get the status for the _current_ user. Otherwise, you are just letting anyone get the access of other user's pets.

Chris-Johnston replied:
So, you shouldn't have to specify the id in the query string at all.

Chris-Johnston commented on PetGame/Controllers/UserController.cs:
```diff
@@ -76,9 +69,9 @@ public IActionResult PostPet([FromForm, FromBody] CreatePetModel model)
                 // HACK: Throw the new pet id into the query string just so we know that an id was added
                 return Redirect($"/?PetId={pet.PetId}");
```
Either this diff is really broken or you are doing something seriously wrong here.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -12,11 +12,22 @@ namespace PetGame.Services
+        const double HungerIncrease = 0.10;
+        const double FeedingIncrease = 0.10;
+        const double TrainingDecrease = 0.15;
+        const double RaceDecrease = 0.20;
```
Can you explain what these do? `RaceDecrease` does not explain very much.

Chris-Johnston commented:
Also, you really need to fix the build status.

anishprasad01 committed:
Moved functions of Pet Status into helper methods

anishprasad01 committed:
Moved more functionality into a separate function

Added comments to const variables

anishprasad01 committed:
Fixed CodeFactor Error

anishprasad01 committed:
Added comments to Get (status) in UserController

anishprasad01 commented:
Ready for (re)review

Chris-Johnston commented on PetGame/Controllers/UserController.cs:
```diff
@@ -52,15 +52,38 @@ public IActionResult GetWhoAmI()
+            //if the user is not logged in, inform them
+            if (u == null)
+            {
+                return Ok("You are not logged in.");
```
We should not return the status code "Ok" if the user is not logged in, we need to return Unauthorized.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,244 @@ public bool DeletePet(ulong petId, ulong ownerId)
+
+                //check the number of minute since the last activity
+                int MinutesSinceLastActivity = 0;
+                MinutesSinceLastActivity = DateTime.Now.Minute - LastActivity.Timestamp.Minute;
```
So, what happens at `12:01` when the last activity was at `11:59`?
I don't think you are comparing dates correctly. You should not just subtract the minute values, but rather determine the length of time between them (perhaps using `Timespan`?)

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,244 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            return HoursSinceLastActivity;
+        }
+
+        private int CalculateMinutesTillNextAction(IEnumerable<Activity> PastActivities, int MinutesTillNextAction, int HoursToCheck)
```
Nitpick, but I do not think that use of `Till` in a name follows good convention.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,244 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            return hungerPercentage;
+        }
+
+        private double ModifyPercentage(double percentage, double changeFactor, char operation)
```
This is exactly the case where you should be using an enum, not a char.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,244 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                return percentage;
+            }
+            //s stands for subtract
+            else if (operation == 's')
```
Also, why can't you just perform the action normally, then call `ConstrainPercentages` on the result?

anishprasad01 replied:
What, like with a pointer?

Chris-Johnston replied:
No like, `var result = Constrain(value + 0.1)`

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,244 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                {
+                    percentage = MinimumPercentage;
+                }
+                return percentage;
```
It really looks like you are doing more work than you have to here.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,244 @@ public bool DeletePet(ulong petId, ulong ownerId)
+        {
+            if (percentage < MinimumPercentage)
+            {
+                percentage = MinimumPercentage;
```
Assignment is pointless in this case, you could just return the value here.

anishprasad01 committed:
Status returns 401 for a user who isn't logged in

Changed the use or chars to determine operation to an enum

Changed time checks to TimeSpans

Misc. Code cleanup

anishprasad01 committed:
Further code cleanup

Added more comments

anishprasad01 committed:
Removed extra number. Replaced it with previously declared constant

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,267 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                    //if a Training event has occured, subtract
+                    else if (Activity.Type.Equals(ActivityType.Training))
+                    {
+                        hungerPercentage = ModifyPercentage(hungerPercentage, TrainingDecrease, PercentageOperation.Subtract);
```
I still don't see why this is as complex as it is, couldn't you just get the same result by doing:
```cs
hungerPercentage = ConstrainPercentage(hungerPercentage - TrainingDecrease);
```

anishprasad01 replied:
Because I thought you wanted everything in separate functions

Chris-Johnston replied:
I think you misunderstood what I meant. A function for `feed` that performs the logic on status when feeding, or a function for `train` that does the same was what I meant, so that in your `switch`/`case` you only call those methods, instead of doing all of the logic inline.

Chris-Johnston replied:
Currently, you are repeating the same logic in more than one place.

anishprasad01 replied:
Gotcha. I think I see what you're getting at with this Modify function being too complex, I'll change it

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,267 @@ public bool DeletePet(ulong petId, ulong ownerId)
+                }
+            }
+            //s stands for subtract
+            else if (operation.Equals(PercentageOperation.Subtract))
```
`enum`s are primitives, they are backed by only primitive types. You do not need to use `Equals()`

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,267 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            //if neither case occurs, (i.e on invalid input), do nothing
+            else
+            {
+                return percentage;
```
Unreachable code, but ok.

Chris-Johnston commented on PetGame/Services/PetService.cs:
```diff
@@ -200,5 +221,267 @@ public bool DeletePet(ulong petId, ulong ownerId)
+            {
+                //percentage is currently greater than the min, and after
+                //the operation, will still be greater than the min
+                if (percentage > 0 && ((percentage -= changeFactor) > MinimumPercentage))
```
You are repeating yourself again here, this is essentially checking the constraints.. like you already do in the constrain method.

anishprasad01 committed:
Removed ModifyPercentages function to simplify logic

Made cooldown length into a const variable

anishprasad01 committed:
Fixed Error where hunger would become negative if no activities had occured within the time span

anishprasad01 committed:
Fixed code issues in GetPetStatusList

Returns empty list instead of null

anishprasad01 committed:
Removed PercentageOptions enum

anishprasad01 committed:
Added the missing PostPet function to the UserController

anishprasad01 commented:
Added PostPet function back.

Chris-Johnston commented:
will merge after CI passes
